dynamic: panic in decodeVarint

Hi, this is a bug I found while fuzzing the library so it's probably pretty unlikely anyone will see this in normal usage.

I've been using the unknown field support in dynamic.Message to write a utility to enrich a MessageDescriptor so that it pretends it knows about all the fields present in a message (i.e. so that unknown fields can be marshalled to JSON (and unmarshalled again)).

While fuzzing my implementation I found this index out of range panic in `github.com/jhump/protoreflect/dynamic.(*codedBuffer).decodeVarint`:

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/jhump/protoreflect/dynamic.(*codedBuffer).decodeVarint(0xc0000fdd90, 0x2, 0x0, 0x0)
	$GOPATH/github.com/jhump/protoreflect/dynamic/codec.go:79 +0x6ad
github.com/jhump/protoreflect/dynamic.(*codedBuffer).decodeTagAndWireType(0xc0000fdd90, 0xbd7fbdebfdef5fef, 0x0, 0x0)
	$GOPATH/github.com/jhump/protoreflect/dynamic/codec.go:172 +0x47
github.com/jhump/protoreflect/dynamic.skipGroup(0xc0000fdd90, 0x0, 0x0, 0x0, 0x0)
	$GOPATH/github.com/jhump/protoreflect/dynamic/binary.go:667 +0x1ba
github.com/jhump/protoreflect/dynamic.(*Message).unmarshalUnknownField(0xc00014b7d0, 0x30000000c, 0xc0000fdd90, 0x0, 0x0)
	$GOPATH/github.com/jhump/protoreflect/dynamic/binary.go:639 +0x3fb
github.com/jhump/protoreflect/dynamic.(*Message).unmarshal(0xc00014b7d0, 0xc0000fdd90, 0x0, 0x0, 0x0)
	$GOPATH/github.com/jhump/protoreflect/dynamic/binary.go:422 +0x203
github.com/jhump/protoreflect/dynamic.(*Message).UnmarshalMerge(0xc00014b7d0, 0x4d70000, 0xc, 0x200000, 0x1, 0x4a50100)
	$GOPATH/github.com/jhump/protoreflect/dynamic/binary.go:403 +0xa7
github.com/jhump/protoreflect/dynamic.(*Message).Unmarshal(0xc00014b7d0, 0x4d70000, 0xc, 0x200000, 0xc00014b7d0, 0xc0000fde01)
	$GOPATH/github.com/jhump/protoreflect/dynamic/binary.go:392 +0x73
github.com/golang/protobuf/proto.Unmarshal(0x4d70000, 0xc, 0x200000, 0x173c440, 0xc00014b7d0, 0x5d2f3018, 0x397abca8)
	$GOPATH/github.com/golang/protobuf/proto/decode.go:340 +0x141
github.com/bradleyjkemp/grpc-tools/internal/proto_decoder.Fuzz(0x4d70000, 0xc, 0x200000, 0x3)
	$GOPATH/github.com/bradleyjkemp/grpc-tools/internal/proto_decoder/fuzz.go:28 +0x265
go-fuzz-dep.Main(0xc0000fdf80, 0x1, 0x1)
	/var/folders/6m/r713l1xj6_dcykbx21svs09h0000gp/T/go-fuzz-build861821362/goroot/src/go-fuzz-dep/main.go:36 +0x1b6
main.main()
	/var/folders/6m/r713l1xj6_dcykbx21svs09h0000gp/T/go-fuzz-build861821362/gopath/src/github.com/bradleyjkemp/grpc-tools/internal/proto_decoder/go.fuzz.main/main.go:15 +0x52
exit status 2
```

You can replicate this using this fuzzing harness with `data = "c\x02���O"`:
```go
package proto_decoder

import (
	"github.com/golang/protobuf/proto"
	"github.com/golang/protobuf/ptypes/empty"
	"github.com/jhump/protoreflect/desc"
	"github.com/jhump/protoreflect/desc/builder"
	"github.com/jhump/protoreflect/dynamic"
)

func Fuzz(data []byte) int {
	d, err := desc.LoadMessageDescriptorForMessage(&empty.Empty{})
	if err != nil {
		panic(err)
	}
	mb, err := builder.FromMessage(d)
	if err != nil {
		panic(err)
	}

	msg, err := mb.Build()
	if err != nil {
		panic(err)
	}

	decoded := dynamic.NewMessage(msg)
	err = proto.Unmarshal(data, decoded)
	if err != nil {
		return 0
	}

	return 1
}
```

--- a/dynamic/binary.go
+++ b/dynamic/binary.go
@@ -690,11 +690,18 @@ func skipGroup(b *codedBuffer) (int, int, error) {
 				}
 				i++
 			}
-			b.index = i + 1
+			// TODO: This would only overflow if buffer length was MaxInt and we
+			// read the last byte. This is not a real/feasible concern on 64-bit
+			// systems. Something to worry about for 32-bit systems? Do we care?
+			b.index++
 		case proto.WireBytes:
 			l, err := b.decodeVarint()
 			if err != nil {
 				return 0, 0, err
+			}
+			lInt := int(l)
+			if lInt < 0 {
+				return 0, 0, fmt.Errorf("proto: bad byte length %d", lInt)
 			}
 			if !b.skip(int(l)) {
 				return 0, 0, io.ErrUnexpectedEOF

--- a/dynamic/codec.go
+++ b/dynamic/codec.go
@@ -36,7 +36,7 @@ func (cb *codedBuffer) eof() bool {
 
 func (cb *codedBuffer) skip(count int) bool {
 	newIndex := cb.index + count
-	if newIndex > len(cb.buf) {
+	if newIndex < cb.index || newIndex > len(cb.buf) {
 		return false
 	}
 	cb.index = newIndex
@@ -258,13 +258,13 @@ func (cb *codedBuffer) decodeRawBytes(alloc bool) (buf []byte, err error) {
 
 	if !alloc {
 		buf = cb.buf[cb.index:end]
-		cb.index += nb
+		cb.index = end
 		return
 	}
 
 	buf = make([]byte, nb)
 	copy(buf, cb.buf[cb.index:])
-	cb.index += nb
+	cb.index = end
 	return
 }
 

